pkg/tls/certificatemanagement/ca.go

```diff
@@ -0,0 +1,592 @@
+// brings their own secrets, CertificateManager will preserve and return them.
+type CertificateManager interface {
+	// CertificateManagement returns the CertificateManagement object or nil if it is not configured.
+	CertificateManagement() *operatorv1.CertificateManagement
+	// GetKeyPair returns an existing KeyPair. If the KeyPair is not found, nil is returned.
+	GetKeyPair(cli client.Client, secretName, secretNamespace string) (KeyPair, error)
+	// GetOrCreateKeyPair returns a KeyPair. If one exists, some checks are performed. Otherwise, a new KeyPair is created.
+	GetOrCreateKeyPair(cli client.Client, secretName, secretNamespace string, dnsNames []string) (KeyPair, error)
+	// GetCertificate returns a Certificate. If the certificate is not found a k8s.io NotFound error is returned.
+	GetCertificate(cli client.Client, secretName, secretNamespace string) (Certificate, error)
+	// Issued returns true if the provided certificate was signed by CertificateManager and returns false for user provided KeyPairs or Certificate Management.
+	Issued(certificate Certificate) bool
+	AddToStatusManager(manager status.StatusManager, namespace string)
+	KeyPair
+}
+
+// KeyPair wraps a Secret object that contains a private key and a certificate. Whether CertificateManagement is
+// configured or not, KeyPair returns the right InitContainer, Volumemount or Volume (when applicable).
+type KeyPair interface {
+	//UseCertificateManagement returns true if this key pair was not user provided and certificate management has been configured.
+	UseCertificateManagement() bool
+	// BYO returns true if this KeyPair was provided by the user. If BYO is true, UseCertificateManagement is false.
+	BYO() bool
+	InitContainer(namespace, csrImage string) corev1.Container
+	VolumeMount(folder string) corev1.VolumeMount
+	Volume() corev1.Volume
+	// HasSkipRenderInOperatorNamespace Default: false. If true, this will avoid (re-)rendering a secret in the operator namespace and avoid changing an ownerRef.
+	HasSkipRenderInOperatorNamespace() bool
+	// SetSkipRenderInOperatorNamespace will avoid (re-)rendering a secret in the operator namespace and avoid changing an ownerRef.
+	SetSkipRenderInOperatorNamespace()
+	Certificate
+}
+
+// Certificate wraps the certificate. Combine this with a TrustedBundle, to mount a trusted certificate bundle to a pod.
+type Certificate interface {
+	X509Certificate() *x509.Certificate
+	Secret(namespace string) *corev1.Secret
+	HashAnnotationKey() string
+	HashAnnotationValue() string
+}
+
+// TrustedBundle is used to create a trusted certificate bundle of the CertificateManager CA and 0 or more Certificates.
+type TrustedBundle interface {
+	MountPath() string
+	ConfigMap(namespace string) *corev1.ConfigMap
+	HashAnnotations() map[string]string
+	VolumeMount() corev1.VolumeMount
+	Volume() corev1.Volume
+	//AddPEM adds a PEM block to the certificate bundle.
+	AddPEM(annotationName string, pem []byte)
```

Comment:
Do we need to have this "Add" functions, or could the pems / certificates be created before the TustedBundle is created and required as part of constructor.

Reply:
This one is added for a one-off case where the typha ca can be brought by a customer. This is a configmap, not a keypair. The only use of this method.

Reply:
I'm not a very big fan of adding it to the constructor. It means that every other place will need to add an extra argument to the constructor that isn't self-explanatory.

Reply:
But you already have the constructor for it, it seems like you've just added another, possibly non intuitive, way of adding a certificate to the trusted bundle. Why not just create this as an implementation of the Certificate interface and add it through the constructor?

The same thing goes for the AddCertificate function. Certificates are required to have a trusted bundle, and you can always created them (at least that's how the uses of this read) before you create the trusted bundle.